### PR TITLE
Improve performance for many entities

### PR DESCRIPTION
**Pull request by Wumpf**
### What

* Fixes https://github.com/rerun-io/rerun/issues/2285
(.. at least fixes it well enough for the moment I reckon)

Improves both the per space view rendering and the space view adding heuristic. In particular on the later there's still significant gains to be made.
Most importantly we should be scaling now much better with number of added systems.

Went fairly early with the decision to have the primary datastructure to pass around be a `Map<System, Vec<Entity>>` (conceptual) instead of the other way round since this is what we need in the systems most of the time.
This is another step towards a stronger contract of systems specifying what components they will query ahead of time!

The way it is implemented on thr (rename from `BlueprintTree` 💥) `SpaceViewContents` also paves the way for ui selection of systems for a given entity path which complements the ongoing work on our new data ingestion interfaces.

-----

Testcase: 
`examples/python/open_photogrammetry_format/main.py --no-frames`, **fully reset viewer** and then hiding **NOT REMOVING** the points (the world/pcl entity) and deselecting them.
(The bold marked pieces are very important as they have an influence on what heuristics run - see #3077)

Before:
image

After:
image

(release runs, averaged over a bunch of frames on my M1 Max)

Highlights:
* `AppState::show`: 24.0ms :arrow_right: 14.6ms 
   * `SpaceViewBlueprint::scene_ui`: 4.9ms :arrow_right: 2.6ms
      * this was the original primary optimization target!
   * `Viewport::on_frame_start`: 15.2ms :arrow_right: 7.4ms
       * `default_created_space_view`: 12.7ms ➡️ 5.1ms
           * quite some time spent on this, but we want to do a paradigm shift there: https://github.com/rerun-io/rerun/issues/3079
           
(slight number discrepancies from the screenshot are due to doing a different run)

----
Draft todo: Code sanity check. Suspecting some heuristics might be slightly broken, need to go through examples

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3078) (if applicable)

- [PR Build Summary](https://build.rerun.io/pr/3078)
- [Docs preview](https://rerun.io/preview/fc5bb0b8908e2e8c666a6d5daa1dee105fe2aae8/docs) 
- [Examples preview](https://rerun.io/preview/fc5bb0b8908e2e8c666a6d5daa1dee105fe2aae8/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: crates/re_viewport/src/space_view.rs [hunk 1]
```diff
@@ -109,17 +124,22 @@ impl SpaceViewBlueprint {
         ctx: &mut ViewerContext<'_>,
         spaces_info: &SpaceInfoCollection,
         view_state: &mut dyn SpaceViewState,
+        entities_per_system: &EntitiesPerSystem,
```
Comment (jleibs) [inline crates/re_viewport/src/space_view.rs:127]: Rename to: "entities_per_system_for_class" or just pass through EntitiesPerSystemPerClass directly.

File: crates/re_viewport/src/space_view.rs [hunk 2]
```diff
@@ -248,9 +268,28 @@ impl SpaceViewBlueprint {
...
         if !entities.is_empty() {
-            self.data_blueprint
+            self.contents
                 .insert_entities_according_to_hierarchy(entities.iter(), &self.space_origin);
             self.entities_determined_by_user = true;
         }
     }
+
+    /// Resets the [`SpaceViewContents::per_system_entities`] for all paths that are part of this space view.
+    pub fn reset_systems_per_entity_path(&mut self, entities_per_system: &EntitiesPerSystem) {
```
Comment (jleibs) [inline crates/re_viewport/src/space_view.rs:278]: Also entities_per_system_for_class

File: crates/re_viewport/src/space_view.rs [hunk 3]
```diff
@@ -248,9 +268,28 @@ impl SpaceViewBlueprint {
...
+        re_tracing::profile_function!();
+
+        let mut per_system_entities = re_viewer_context::PerSystemEntities::new();
```
Comment (jleibs) [inline crates/re_viewport/src/space_view.rs:281]:
```suggestion
        // TODO(): We believe this is *correct* but not necessarily optimal. Pay attention
        // to the algorithmic complexity here as we consider changing the indexing and
        // access patterns of these structures in the future.
        let mut per_system_entities = re_viewer_context::PerSystemEntities::new();
```

File: crates/re_space_view/src/space_view_contents.rs [hunk 4]
```diff
@@ -95,8 +95,7 @@ pub struct DataBlueprintTree {
     /// Two things to keep in sync:
     /// * children on [`DataBlueprintGroup`] this is on
     /// * elements in [`Self::path_to_group`]
-    /// TODO(andreas): Can we reduce the amount of these dependencies?
-    entity_paths: IntSet<EntityPath>,
+    per_system_entity_list: PerSystemEntities,
```
Comment (jleibs) [inline crates/re_space_view/src/space_view_contents.rs:98]: This should be skip serde for now (until it's totally refactored anyways when we stop using serde-field, etc.)

File: crates/re_space_view/src/space_view_contents.rs [hunk 5]
```diff
@@ -270,7 +304,14 @@ impl DataBlueprintTree {
         let mut new_leaf_groups = Vec::new();
 
         for path in paths {
-            self.entity_paths.insert(path.clone());
+            // TODO(andreas/jleibs): Incoming entities should already "know" what system they belong to.
```
Comment (jleibs) [inline crates/re_space_view/src/space_view_contents.rs:307]: This doesn't need to happen at all -- we're going to recompute it based on the group hierarchy on every frame at which point it won't be unknown.

**Review by jleibs — APPROVED** [review]

**Comment by emilk** [comment]
`cargo r -p raw_mesh` is also a pretty good benchmark

File: crates/re_viewport/src/space_view_heuristics.rs [hunk 6]
```diff
@@ -393,3 +431,96 @@ fn is_entity_processed_by_part_collection(
 
     false
 }
+
+pub fn default_entities_per_system_per_class(
```
[the remark this answers is not in the document]
Comment (Wumpf) [reply crates/re_viewport/src/space_view_heuristics.rs:435]: was about to reply that it's the same `default` as in `default_created_space_views`. But that's not accurate, just `entities_per_system_per_class` is more appropriate

**Pull request merged**